website/marietje/api/v1/views.py

```diff
@@ -0,0 +1,218 @@
+from marietje.api.v1.serializers import PlayerSerializer, QueueItemSerializer
+from marietje.models import Player, SpotifyQueueItem
+from tosti.api.openapi import CustomAutoSchema
+
+
+class PlayerListAPIView(ListAPIView):
+    """
+    Player List API View.
+
+    Permissions required: None
+
+    Use this endpoint to get a list of all installed Spotify Players.
+    """
+
+    serializer_class = PlayerSerializer
+    queryset = Player.objects.all()
+
+
+class PlayerRetrieveAPIView(RetrieveAPIView):
+    """
+    Player Retrieve API View.
+
+    Permissions required: None
+
+    Use this endpoint to get the details of an installed Spotify Players.
+    """
+
+    serializer_class = PlayerSerializer
+    queryset = Player.objects.all()
+
+
+class PlayerQueueListAPIView(ListAPIView):
+    """
+    Player Queue List API View.
+
+    Permissions required: None
+
+    Use this endpoint to get a list of tracks that are in the queue of the Spotify Player. Tracks are sorted on the
+    time that they were added, the track that was added last will appear first in the list.
+    """
+
+    serializer_class = QueueItemSerializer
+    queryset = SpotifyQueueItem.objects.all()
+    pagination_class = StandardResultsSetPagination
+
+    def get_queryset(self):
+        """Get the queryset."""
+        return self.queryset.filter(player=self.kwargs.get("player"))
+
+
+class PlayerTrackSearchAPIView(APIView):
+    """Player Track Search API View."""
+
+    schema = CustomAutoSchema(
+        manual_operations=[{"name": "query", "in": "query", "required": True, "schema": {"type": "string"}}],
+        response_schema={
+            "type": "object",
+            "properties": {
+                "query": {"type": "string", "example": "string"},
+                "id": {"type": "int", "example": "123"},
+                "results": {
+                    "type": "array",
+                    "items": {
+                        "type": "object",
+                        "properties": {
+                            "name": {"type": "string", "example": "string"},
+                            "artists": {"type": "array", "items": {"type": "string", "example": "string"}},
+                            "id": {"type": "string", "example": "string"},
+                        },
+                    },
+                },
+            },
+        },
+    )
+
+    def get(self, request, **kwargs):
+        """
+        Search for a Spotify track.
+
+        Permission required: marietje.can_request
+
+        Use this endpoint to search for a Spotify Track. Tracks can be searched via their Spotify id.
+        """
+        player = kwargs.get("player")
+        query = request.GET.get("query", "")
+        request_id = request.GET.get("id", None)
+        try:
+            maximum = int(request.GET.get("maximum", 5))
+        except ValueError:
+            maximum = 5
+        if request.user.has_perm("marietje.can_request", player):
```

Comment:
APIView has a dedicated `check_permissions` / `check_object_permissions` method for this which I think we should use (see http://www.cdrf.co/3.9/rest_framework.views/APIView.html)